scylla/src/frame/types.rs

```diff
@@ -525,3 +525,690 @@ fn type_inet() {
     let read_iv6 = read_inet(&mut &*buf).unwrap();
     assert_eq!(iv6, read_iv6);
 }
+
+fn zig_zag_encode(v: i64) -> u64 {
+    ((v >> 63) ^ (v << 1)) as u64
+}
+
+fn zig_zag_decode(v: u64) -> i64 {
+    ((v >> 1) as i64) ^ -((v & 1) as i64)
+}
+
+fn unsigned_vint_encode(v: u64, buf: &mut Vec<u8>) {
+    let mut v = v;
+    let mut number_of_bytes = (639 - 9 * v.leading_zeros()) >> 6;
+    if number_of_bytes <= 1 {
+        return buf.put_u8(v as u8);
+    }
+
+    if number_of_bytes != 9 {
+        let extra_bytes = number_of_bytes - 1;
+        let length_bits = !(0xff >> extra_bytes);
+        v |= (length_bits as u64) << (8 * extra_bytes);
+    } else {
+        buf.put_u8(0xff);
+        number_of_bytes -= 1;
+    }
+    buf.put_uint(v, number_of_bytes as usize)
+}
+
+fn unsigned_vint_decode(buf: &mut &[u8]) -> Result<u64, ParseError> {
+    let first_byte = buf.read_u8()?;
+    let extra_bytes = first_byte.leading_ones() as usize;
+
+    let mut v = if extra_bytes != 8 {
+        let first_byte_bits = first_byte & (0xffu8 >> extra_bytes);
+        (first_byte_bits as u64) << (8 * extra_bytes)
+    } else {
+        0
+    };
+
+    if extra_bytes != 0 {
+        v += buf.read_uint::<BigEndian>(extra_bytes)?;
+    }
+
+    Ok(v)
+}
+
+pub(crate) fn vint_encode(v: i64, buf: &mut Vec<u8>) {
+    unsigned_vint_encode(zig_zag_encode(v), buf)
+}
+
+pub(crate) fn vint_decode(buf: &mut &[u8]) -> Result<i64, ParseError> {
+    unsigned_vint_decode(buf).map(zig_zag_decode)
+}
+
+#[test]
+fn zig_zag_encode_test() {
+    assert_eq!(zig_zag_encode(0), 0);
+    assert_eq!(zig_zag_encode(-1), 1);
+    assert_eq!(zig_zag_encode(1), 2);
+    assert_eq!(zig_zag_encode(-2), 3);
+    assert_eq!(zig_zag_encode(2), 4);
+    assert_eq!(zig_zag_encode(-3), 5);
+    assert_eq!(zig_zag_encode(3), 6);
+}
+
+#[test]
+fn zig_zag_decode_test() {
+    assert_eq!(zig_zag_decode(0), 0);
+    assert_eq!(zig_zag_decode(1), -1);
+    assert_eq!(zig_zag_decode(2), 1);
+    assert_eq!(zig_zag_decode(3), -2);
+    assert_eq!(zig_zag_decode(4), 2);
+    assert_eq!(zig_zag_decode(5), -3);
+    assert_eq!(zig_zag_decode(6), 3);
+}
+
+#[test]
+fn unsigned_vint_encode_and_decode_test() {
+    let unsigned_vint_encoding = vec![
```

Comment:
Could we auto-generate at least some of those test cases?

Reply:
But the encodings need to be hardcoded anyway, don't they? There's almost nothing to gain in terms of LoCs.

How these cases were generated:
1. Python script generated test cases in Go
2. Go program with vint encoding logic copied from Go Driver took generated test cases and printed the encodings.
3. Python script took the output and generated Rust test cases.

Reply:
OK, it makes sense to keep them spelled out. Let's keep them. The tests cases in `vint_encode_and_decode_test` should definitely be autogenerated, though.